### DISCUSSION TITLE
How to easily reblame on previous version of a commit?

### DISCUSSION THREAD

**Discussion by mux** [tpope/vim-fugitive#2274] [General]
One thing that happens a lot to me when I use :Git blame is that I locate the commit where a line has last changed, but it turns out that that change was only fixing the style/indentation. I haven't found a way to use the fugitive plugin to easily reblame on previous revision of a file. I initially thought that the '-' key would help me do that, but either I am not using it correctly, or it does not actually do what I thought it would.

Is there anyway to support this kind of functionality so we can easily walk our way back the revision tree using blame?

**Comment by tpope** [comment]
This is a common question. Please help me understand why so I can better document it.

> it turns out that that change was only fixing the style/indentation

The blame window only shows the SHA, the author, and the date. What is the implied next step that leads you to the conclusion that it was a style/indentation change?

**Reply by mux** [reply]
Apologies, I should have been clearer. When I am in the blame window, I just hit enter on the relevant line, and fugitive opens up a new window showing me the diff. This is when I realize that the change was a whitespace change, and when I wish I could just say "reblame on the commit preceding the commit I am currently viewing".

**Reply by tpope** [reply]
When you're in the commit, you can hit enter again on the old version of the line, which starts with a `-`, and that will take you to the file/line in the commit preceding the commit you are currently viewing. Then you can press enter again to reblame.

This is in the docs, but I guess needs to be made more prominent: https://github.com/tpope/vim-fugitive/blob/41beedabc7e948c787ea5696e04c3544c3674e23/doc/fugitive.txt#L75-L83

**Reply by mux** [reply]
I am sorry - I was sure that I had been reading the documentation fully and failed to see anything related to that, but now I see that I can hit enter on the diff to see the file at the revision before that change, and then type ":Git blame" again, at least if I understood correctly. I think there would be value in making this process shorter however.

**Reply by mux** [reply]
I must be doing something wrong, but hitting enter again does not seem to result in a reblame, it opens the revision of the file prior to the commit I was viewing before. This is nice because I can do :Git blame again, but does not seem to match the documentation. Am I missing something?

**Reply by mux** [reply]
Actually, it takes me back to the revision after the commit, so typing :Git blame again doesn't work.

**Reply by mux** [reply]
So sorry for the back and forth. Retyping :Git blame does work as long as I take care of hitting enter on the old line, starting with '-' just as you said. I still cannot figure out how to get a reblame though, as I was saying, hitting enter again just shows me the revision of the file, not the reblame.

**Reply by tpope** [reply]
> I must be doing something wrong, but hitting enter again does not seem to result in a reblame, it opens the revision of the file prior to the commit I was viewing before

Pressing enter on `-` takes you to the file. Pressing enter _again_ reblames.

> I think there would be value in making this process shorter however.

There's maps that can make finding the `-` line easier:

https://github.com/tpope/vim-fugitive/blob/41beedabc7e948c787ea5696e04c3544c3674e23/doc/fugitive.txt#L429-L435

I don't see it getting much shorter than that, but do go on.

**Reply by mux** [reply]
No matter how much I press enter I do not get a reblame. Here's what I see precisely:

- I use :Git blame and get the new window with the blame information.
- In this window, I press enter on the relevant line, and get a window with the relevant diff.
- In the diff window, I press enter on the "old" line (starting with -), and it opens up that "old" revision, before the patch I was just watching.
- I press enter again, and then instead of getting a reblame (if I understand correctly), I get the diff that led to the line as it was in the old revision. It's as if it skipped a a step: it doesn't reblame, but does show me the diff I would see if I had reblamed and pressed enter on that line again.

I suppose this is where the confusion comes from. And I suppose that you thought there would be little point in showing the actual blame on that "old" revision if we're only interested in that line we've been pressing enter on, so it made more sense to just show the previous diff relevant to that line. And you might very well be right, the only thing that ends up being very slightly annoying to me is that I do not see the author right away. I have to scroll back up in that diff to see who the author is, which is really the information I was looking for. Arguably however, if it did show the reblame, I would end up having to press enter a third time to check the diff anyways - but I would already know the author.

So this is all very much subjective I guess, but at least I wanted to make it clear where the confusion comes from. I kind of feel that adding the extra step I am referring to would actually be better, but there isn't much else I can say to substantiate this.

PS: let me just say thank you for this amazing git plugin and all the amazing other git plugins you created. I don't want this discussion to just be about vague complaints :-)

**Reply by tpope** [reply]
You're right, I did forget it takes it a step further.

> The only thing that ends up being very slightly annoying to me is that I do not see the author right away. I have to scroll back up in that diff to see who the author is, which is really the information I was looking for. Arguably however, if it did show the reblame, I would end up having to press enter a third time to check the diff anyways - but I would already know the author.

Now the disconnect is I don't understand why the author is important. Commit message, sure, I scroll up for that all the time. But author? That's almost never actionable to me on its own. I wouldn't want to press enter an extra time _just_ to see the author.

**Reply by mux** [reply]
> Now the disconnect is I don't understand why the author is important. Commit message, sure, I scroll up for that all the time. But author? That's almost never actionable to me on its own. I wouldn't want to press enter an extra time _just_ to see the author.

Well then how do you blame the person responsible? :-) Seriously though, I suppose it depends entirely on the context. In my line of work, I do often need to know who authored a particular line of code and potentially talk to them, if only because their intention is often not clear. I guess what is most important here is that if the documentation says "reblame", I would expect to end up with a window that is similar to the one I get when I type :Git blame, but then that is just a minor documentation issue.

**Reply by tpope** [reply]
> Well then how do you blame the person responsible? :-) Seriously though, I suppose it depends entirely on the context. In my line of work, I do often need to know who authored a particular line of code and potentially talk to them, if only because their intention is often not clear.

Sure, but you checked the intention wasn't clearly stated in the commit message before bothering them, right? Right?

> I guess what is most important here is that if the documentation says "reblame", I would expect to end up with a window that is similar to the one I get when I type :Git blame, but then that is just a minor documentation issue.

It's trying to convey this with "reblame **at that line**", but that's not at all clear.

**Reply by mux** [reply]
You are absolutely right - I didn't have the correct process in my head. If intention is not clear it must mean that I already read the commit log. Sometimes though, intent is clear but the code is just bad and I still need to talk with people. :-) But like I was saying, this is really super minor and I am completely fine with the way the plugin works today.
